StandaloneMmPkg/Drivers/StandaloneMmIplPei/StandaloneMmIplPei.c

```diff
@@ -217,6 +217,134 @@ LocateMmFvForMmCore (
+
+  @retval HobList              If fundation and platform HOBs not existed,
+                               it is pointed to PEI HOB List. If existed,
+                               it is pointed to fundation and platform HOB list.
+**/
+VOID *
+CreatMmHobList (
+  OUT UINTN                           *HobSize,
+  IN  MM_COMM_BUFFER                  *MmCommBuffer,
+  IN  EFI_PHYSICAL_ADDRESS            MmFvBase,
+  IN  UINT64                          MmFvSize,
+  IN  EFI_GUID                        *MmCoreFileName,
+  IN  PHYSICAL_ADDRESS                MmCoreImageAddress,
+  IN  UINT64                          MmCoreImageSize,
+  IN  PHYSICAL_ADDRESS                MmCoreEntryPoint,
+  IN  EFI_MMRAM_HOB_DESCRIPTOR_BLOCK  *Block
+  )
+{
+  EFI_STATUS  Status;
+  VOID        *HobList;
+  VOID        *PlatformHobList;
+  UINTN       PlatformHobSize;
+  UINTN       BufferSize;
+  UINTN       FoundationHobSize;
+
+  //
+  // Get platform HOBs
+  //
+  PlatformHobSize = 0;
+  Status          = CreateMmPlatformHob (NULL, &PlatformHobSize);
+  if (Status == RETURN_BUFFER_TOO_SMALL) {
+    ASSERT (PlatformHobSize != 0);
+    //
+    // Creat platform HOBs for MM foundation to get MMIO HOB data.
+    //
+    PlatformHobList = AllocatePages (EFI_SIZE_TO_PAGES (PlatformHobSize));
+    ASSERT (PlatformHobList != NULL);
+    if (PlatformHobList == NULL) {
+      DEBUG ((DEBUG_ERROR, "%a: Out of resource to create platform MM HOBs\n", __func__));
+      CpuDeadLoop ();
+    }
+
+    BufferSize = PlatformHobSize;
+    Status     = CreateMmPlatformHob (PlatformHobList, &PlatformHobSize);
+    ASSERT_EFI_ERROR (Status);
+    ASSERT (BufferSize == PlatformHobSize);
+  }
+
+  ASSERT_EFI_ERROR (Status);
+
+  //
+  // Get size of foundation HOBs
+  //
+  FoundationHobSize = 0;
+  Status            = CreateMmFoundationHobList (
+                        NULL,
+                        &FoundationHobSize,
+                        PlatformHobList,
+                        PlatformHobSize,
+                        MmCommBuffer,
+                        MmFvBase,
+                        MmFvSize,
+                        MmCoreFileName,
+                        MmCoreImageAddress,
+                        MmCoreImageSize,
+                        MmCoreEntryPoint,
+                        Block
+                        );
+  FreePages (PlatformHobList, EFI_SIZE_TO_PAGES (PlatformHobSize));
+  ASSERT (Status == RETURN_BUFFER_TOO_SMALL);
+  ASSERT (FoundationHobSize != 0);
+
+  //
+  // Final result includes platform HOBs, foundation HOBs and a END node.
+  //
+  *HobSize = PlatformHobSize + FoundationHobSize + sizeof (EFI_HOB_GENERIC_HEADER);
+  HobList  = AllocatePages (EFI_SIZE_TO_PAGES (*HobSize));
+  ASSERT (HobList != NULL);
+  if (HobList == NULL) {
+    DEBUG ((DEBUG_ERROR, "Out of resource to create MM HOBs\n"));
+    CpuDeadLoop ();
+  }
+
+  //
+  // Get platform HOBs
+  //
+  Status = CreateMmPlatformHob (HobList, &PlatformHobSize);

```

Comment:
why do we need to build platform hob for twice? (The first build is at line 275)

Reply:
The 1st platform hob is for MmIplBuildResourceHobForAllSystemMemory and will be released.